warn/warn_docstring.go

```diff
@@ -0,0 +1,250 @@
+func countLeadingSpaces(s string) int {
+	spaces := 0
+	for _, c := range s {
+		if c == ' ' {
+			spaces++
+		} else {
+			break
+		}
+	}
+	return spaces
+}
+
+var argHeaderRegex = regexp.MustCompile("^ *(\\*?\\*?\\w+)( *\\([\\w\\ ,]+\\))?:")
+var argRegex = regexp.MustCompile("[\\*\\w]+")
+
+// parseFunctionDocstring parses a function docstring and returns a docstringInfo object containing
+// the parsed information about the function, its arguments and its return value
+func parseFunctionDocstring(doc *build.StringExpr) docstringInfo {
+	start, _ := doc.Span()
+	indent := start.LineRune - 1
+	lines := strings.Split(doc.Value, "\n")
+
+	// Dedent the lines
+	for i, line := range lines {
+		line = strings.TrimRight(line, " ")
+		if i != 0 {
+			for j, chr := range line {
+				if j >= indent || chr != ' ' {
+					line = line[j:]
+					break
+				}
+			}
+		}
+		lines[i] = line
+	}
+
+	// Split by empty lines
+	blocks := []docstringBlock{}
+	newBlock := true
+	for i, line := range lines {
+		if len(line) == 0 {
+			newBlock = true
+			continue
+		}
+		if newBlock {
+			newBlock = false
+			blocks = append(blocks, docstringBlock{
+				startLineNo: start.Line + i,
+				lines:       []string{},
+			})
+		}
+		blocks[len(blocks)-1].lines = append(blocks[len(blocks)-1].lines, line)
+	}
+
+	info := docstringInfo{}
+	info.args = make(map[string]bool)
+
+	if len(blocks) > 0 && len(blocks[0].lines) == 1 {
+		// Exactly one line in the first block
+		info.hasHeader = true
+	}
+
+	// Iterate over the blocks, extract data
+	for _, block := range blocks {
+		switch block.lines[0] {
+		case "Args:", "Arguments:":
+			if block.lines[0] == "Arguments:" {
+				info.argumentsLine = block.startLineNo
+			}
+
+			argIndentation := -1
+			for _, line := range block.lines[1:] {
+				newIndentation := countLeadingSpaces(line)
+				if newIndentation <= argIndentation || argIndentation == -1 {
+					header := argHeaderRegex.FindString(line)
+					if len(header) > 0 {
+						arg := argRegex.FindString(header)
+						info.args[arg] = true
+						argIndentation = newIndentation
+					}
+				}
+			}
+		case "Returns:":
+			if len(block.lines) > 1 {
+				info.returns = true
+			}
+		}
+	}
+	return info
+}
+
+func getParamName(param build.Expr) string {
+	switch param := param.(type) {
+	case *build.Ident:
+		return param.Name
+	case *build.BinaryExpr:
+		// keyword parameter
+		if ident, ok := param.X.(*build.Ident); ok {
+			return ident.Name
+		}
+	case *build.UnaryExpr:
+		// *args or **kwargs
+		if ident, ok := param.X.(*build.Ident); ok {
+			return param.Op + ident.Name
+		}
+	}
+	return ""
+}
+
+func hasReturnValues(def *build.DefStmt) bool {
+	result := false
+	build.Walk(def, func(expr build.Expr, stack []build.Expr) {
+		ret, ok := expr.(*build.ReturnStmt)
+		if ok && ret.Result != nil {
+			result = true
+		}
+	})
+	return result
+}
+
+func functionDocstringWarning(f *build.File, fix bool) []*Finding {
```

Comment:
opt. The function is quite long. It may be useful to split it.

Reply:
The method almost only contains the emission of warnings, all the parsing logic has already been moved out. I'm not sure how I can split it further.